ext/ddtrace_profiling_native_extension/collectors_dynamic_sampling_rate.c

```diff
@@ -77,16 +92,40 @@ uint64_t dynamic_sampling_rate_get_sleep(dynamic_sampling_rate_state *state, lon
 }
 
 bool dynamic_sampling_rate_should_sample(dynamic_sampling_rate_state *state, long wall_time_ns_before_sample) {
+  long latest_tick_time_ns = long_max_of(0, wall_time_ns_before_sample - state->last_check_time_ns);
+  state->tick_time_ns = ((unsigned long) (EMA_SMOOTHING_FACTOR * latest_tick_time_ns) + ((1.0 - EMA_SMOOTHING_FACTOR) * state->tick_time_ns));
+  state->last_check_time_ns = wall_time_ns_before_sample;
   return wall_time_ns_before_sample >= atomic_load(&state->next_sample_after_monotonic_wall_time_ns);
 }
 
-void dynamic_sampling_rate_after_sample(dynamic_sampling_rate_state *state, long wall_time_ns_after_sample, uint64_t sampling_time_ns) {
+static void dynamic_sampling_rate_after_sample(dynamic_sampling_rate_state *state, long wall_time_ns_after_sample, uint64_t tick_time_ns, uint64_t sampling_time_ns) {
   double overhead_target = state->overhead_target_percentage;
 
   // The idea here is that we're targeting a maximum % of wall-time spent sampling.
-  // So for instance, if sampling_time_ns is 2% of the time we spend working, how much is the 98% we should spend
-  // sleeping? As an example, if the last sample took 1ms and the target overhead is 2%, we should sleep for 49ms.
-  uint64_t time_to_sleep_ns = sampling_time_ns * ((100.0 - overhead_target)/overhead_target);
+  // We have 4 variables:
+  // * sampling_time -> How long did we spend sampling
+  // * overhead_target -> Percentage of time we want sampling_time to represent in relation to total time
+  // * sleeping_time -> How long we want to delay sampling for to keep to overhead_target
+  // * tick_time -> Time between sampling opportunities (0 for continuous operation, time between sampling decisions in discrete ones)
+  // Thus, total_time can be understood to be sampling_time + sleeping_time + tick_time and we want to solve for sleep_time in the
+  // following relation:
+  //
+  //   sampling_time  ----- overhead_target
+  //   total_time     ------ 100%
+  //
+  // Which wields:
+  //
+  //     total_time = 100 * sampling_time / overhead_target <=>
+  // <=> sleeping_time + sampling_time + tick_time  = 100 * sampling_time / overhead_target <=>
+  // <=> sleeping_time = 100 * sampling_time / overhead_target - sampling_time - tick_time
+  //
+  // For a concrete example of continuous sampling where:
+  // * overhead_target = 2%
+  // * sampling_time = 1ms
+  // * between_time = 0
+  //
+  // Then sleeping_time would wield (100 * 1ms) / 2 - 1 = 49ms
```

Comment:
Minor: I suggest using `time_to_sleep` here or switching the equation to use `sleeping_time`, it's weird to use two names for the same thing :)